### ISSUE TITLE
Document Private Registry Authentication

### ISSUE THREAD

**Issue by pixie79** [GoogleCloudPlatform/kubernetes#499]
How do we set up kubernetes to works with private registries?

Currently we can put a remote registry into the image tag for the pod but there appears to be nowhere to enter the login details if these images are private.

Ideally I guess we should be able to set these in the global config script, as the login details for a single registry would be the same for each pod, rather than repeating them in each pod's spec.

**Comment by bgrant0607** [comment]
@erictune @derekwaynecarr @pmorie What else needs to be done on this?

**Comment by smarterclayton** [comment]
As per our discussion at the face to face, one option would be to deliver a secret type that would convey authorization to pull the image.  The kubelet would need to know something about that secret type, and have a way of defending against malicious use of the secret.

**Comment by pmorie** [comment]
Do we want to do pulls in a user container as a long-term goal?

**Comment by smarterclayton** [comment]
I don't know.  It seems to me that the goal of the platform is to abstract image pulling - it's not something a user is concerned with.  It should be fast and ideally delivered by an efficient network abstraction.  Cgroup constraints on pulls in some cases - sure.  Special pullers - maybe.  

Tim mentioned something last week that has also come up from the network filesystems guys - having a mount already established on disk of images so that you don't even need to download anything.  In that model we would need to still check some permissions, and the user container wouldn't play into it too much.

**Comment by bgrant0607** [comment]
+1 to remote mounting of images. That's the only way we're going to get to ~instantaneous container start.

Pulls (and builds) that remain will need to be constrained, however.

**Comment by pmorie** [comment]
+1 also to remote mounting; we'll need to distribute secrets for whatever
ultimately dies the pull for private registries

**Comment by smarterclayton** [comment]
A bunch of folks are about to start working on this on our end - I don't know who will own the proposal but expect one soon.

**Comment by hobti01** [comment]
@smarterclayton Did a proposal come from this issue?

**Comment by smarterclayton** [comment]
@liggitt is working on service accounts right now - his plan was to follow that up with being able to pull images with the secrets associated with the service account.

**Comment by vially** [comment]
Any update on this?

**Comment by liggitt** [comment]
1. Pods now have an `imagePullSecrets` field that lists the secrets to use for pulling the container images
2. That field must reference secrets of type "kubernetes.io/dockercfg", with a ".dockercfg" key containing dockercfg file credentials.
3. If a pod references a service account, that service account's imagePullSecrets are automatically added to the pod's imagePullSecrets field.

**Comment by liggitt** [comment]
@deads2k any pointers to docs or better explanations?

**Comment by bgrant0607** [comment]
We should ensure this is well documented. We have a tiny amount of documentation here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/1f952b05d3e56cf233c94f1b2c65f05e6cde5061/docs/images.md#using-a-private-registry

**Comment by erictune** [comment]
@bgrant0607 what else do you want to see in that documentation.

**Comment by bgrant0607** [comment]
On second glance, the coverage seems fine, actually.

It wasn't clear to me that the bullet list was a table of contents for the succeeding sections. Also, I'd describe imagePullSecrets before either copying .dockercfg to nodes or pre-pulling (that seems like a niche case). And it's not "ImagePullKeys" -- it's "imagePullSecrets".

If we add a convenience command for bundling secrets, we should document that there, also. @liggitt

**Comment by bgrant0607** [comment]
I think this is done.

**Comment by mattma** [comment]
Following the documentation, I have successfully added a `secret`

```
 ➜ kubectl get secrets                                                                                                                                                                                                                                       ✹
NAME            TYPE                      DATA
myregistrykey   kubernetes.io/dockercfg   1
```

My question, all example of using `imagePullSecrets` is in the `POD`, how can i use it in the `RC`? E.G: this is my `xxx-rc.yaml`

```
apiVersion: v1
kind: ReplicationController
metadata:
  labels:
    name: $NAME
    namespace: default
  name: $NAME
spec:
  replicas: 2
  selector:
    name: $NAME
    version: $NAME
  template:
    metadata:
      labels:
        name: $NAME
        version: $NAME
    spec:
      containers:
      - image: $PRIVATE_REPO:tag
        name:$NAME
        ports:
        - containerPort: 8000
      imagePullSecrets:
         - name: myregistrykey
```

**Comment by smarterclayton** [comment]
It works in the RC the same way - just make sure it's in the right place in
the pod template.

On Aug 11, 2015, at 7:12 PM, Matt Ma <notifications@github.com> wrote:

Following the documentation, I have successfully added a secret

 ➜ kubectl get secrets


                                        ✹
NAME            TYPE                      DATA
myregistrykey   kubernetes.io/dockercfg   1

My question, all example of using imagePullSecrets is in the POD, how can i
use it in the RC? E.G: this is my xxx-rc.yaml

apiVersion: v1
kind: ReplicationController
metadata:
  labels:
    name: $NAME
    namespace: default
  name: $NAME
spec:
  replicas: 2
  selector:
    name: $NAME
    version: $NAME
  template:
    metadata:
      labels:
        name: $NAME
        version: $NAME
    spec:
      containers:
      - image: $PRIVATE_REPO:tag
        name:$NAME
        ports:
        - containerPort: 8000
      imagePullSecrets:
         - name: myregistrykey

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/499#issuecomment-130107004>
.

**Comment by mattma** [comment]
@smarterclayton With my setting above, it did not work. 

```
 ➜ kubectl get po                                                                                                                                                                                                                                            ✹
NAME             READY     STATUS                                      RESTARTS   AGE
kube-dns-5ny1k   3/3       Running                                     0          1h
xx-pdl0j         0/1       Error: image xxx:xxx not found   0          37m
xx-zj2rf         0/1       Error: image xxx:xxx not found   0          37m
```

Could you help me what is the right place in `rc`?

**Comment by smarterclayton** [comment]
It's possible it also needs to be added to the service account, @deads2k?

On Aug 11, 2015, at 8:02 PM, Matt Ma <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> With my setting above,
it did not work.

 ➜ kubectl get po


                                        ✹
NAME             READY     STATUS
RESTARTS   AGE
kube-dns-5ny1k   3/3       Running
0          1h
xx-pdl0j         0/1       Error: image xxx:xxx not found   0          37m
xx-zj2rf         0/1       Error: image xxx:xxx not found   0          37m

Could you help me what is the right place in rc?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/499#issuecomment-130116156>
.

**Comment by mattma** [comment]
[docs](http://kubernetes.io/v1.0/docs/user-guide/service-accounts.html) is not clear what to do to fix this use case. 

On the other hand, [docs](http://kubernetes.io/v1.0/docs/user-guide/images.html) explains very clearly to address the issue of Use a hosted private Docker registry. Is this a bug?

**Comment by deads2k** [comment]
> It's possible it also needs to be added to the service account, @deads2k?

`ImagePullSecrets` in a podspec are not gated by the list of secrets on the service account.  What you specify is what gets used.  I don't think we have good logging around exactly which secret is attempted from the keyring, but that might be the next step.

**Comment by erictune** [comment]
@mattma I think it is possible that the pod template and RC aren't being created the way you think they are.  Would you please follow the steps in https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/user-guide/application-troubleshooting.md#my-pod-is-running-but-not-doing-what-i-told-it-to-do and see if that fixes it?  If that still doesn't help, please post the exact RC yaml or json that you are trying to create (anonymize if necessary).

**Comment by mattma** [comment]
@erictune Thank you for the tips. I will give it a try.

I have started a [open source project](https://github.com/mattma/coreos-distro) to launch `kubernetes` with `fleet` `unit-files`. If you follow the README section, you should have a `kubernete` cluster setup in one minute. 

I currently have an [issue or question](https://github.com/mattma/coreos-distro/issues/6) whether Docker landed in `flannel` network or not. It would be great to get some advices from you. Thanks. 

Here is [Master machine Docker Unit file](https://github.com/mattma/coreos-distro/blob/master/setup/cloud-init/master-data#L49)

Here is [Node machine Docker Unit file](https://github.com/mattma/coreos-distro/blob/master/setup/cloud-init/node-data#L51)

**Comment by mattma** [comment]
```
Containers:
  ts:
    Image:		xxx/xxx
    State:		Waiting
      Reason:		Image: xxx/xxx is not ready on the node
    Ready:		False
    Restart Count:	0
Conditions:
  Type		Status
  Ready 	False
Events:
  FirstSeen				LastSeen			Count	From			SubobjectPath	Reason		Message
  Thu, 13 Aug 2015 11:16:05 -0700	Thu, 13 Aug 2015 11:16:05 -0700	1	{scheduler }				scheduled	Successfully assigned xxx-xxx to 172.17.8.101
  Thu, 13 Aug 2015 11:16:05 -0700	Thu, 13 Aug 2015 11:17:16 -0700	8	{kubelet 172.17.8.101}			failedSync	Error syncing pod, skipping: secrets "myregistrykey" not found
```

My `rc` file is posted above which is the full version. I am sure it is related to the `secret`. How do I add a `service account`?

**Comment by erictune** [comment]
@mattma adding a service account won't fix the problem, because you already have the `imagePullSecret` field on your pod.

Two things that might be going on:

 You might be using the new dockercfg format, which we don't support yet.  See this issue for details: https://github.com/kubernetes/kubernetes/issues/12626
@deads2k does #12626 affect imagePullSecrets, in terms of not supporting the new dockercfg format?

- you might have bad data in your secret.  do this:
```
kubectl get secrets myregistrykey -o yaml | grep dockercfg: | cut -f 2 -d : | base64 -D > actual.dockercfg
```
and make sure that file matches what you are expecting.

**Comment by deads2k** [comment]
`secrets "myregistrykey" not found` means that the call to `/api/v1/namespaces/<pod-namespace>/secrets/myregistrykey`, returned a 404.  Since you explicitly set the secret, the kubelet is currently coded to not even attempt the image pull.  https://github.com/kubernetes/kubernetes/pull/12736 relaxes the behavior, but without that image pull secret, you probably won't progress.

I don't think #12626 applies in this case since the secret wasn't found at all.  We don't attempt to parse until the keyring is created.

**Comment by deads2k** [comment]
For reference, I would expect an unmarshalling failure due to the new dockercfg format would fail here: https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L282, but the not found comes from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1290

**Comment by mattma** [comment]
@erictune @deads2k 

```
 ➜ kubectl get secrets myregistrykey -o yaml | grep dockercfg: | cut -f 2 -d : | base64 -D > actual.dockercfg                       ↑ ✹ ✭
zsh: file exists: actual.dockercfg
Error from server: secrets "myregistrykey" not found

 ➜ curl http://172.xx.xx.x:8080/api/v1/namespaces/default/secrets/myregistrykey                                                 1 ↵↑ ✹ ✭
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "secrets \"myregistrykey\" not found",
  "reason": "NotFound",
  "details": {
    "name": "myregistrykey",
    "kind": "secrets"
  },
  "code": 404
}%
```

**Comment by drora** [comment]
How is it working for you? no matter which approach I try, it fails to pull from my private repo.
- minions with /.dockercfg didn't work. tried all suggested permutations/location. 
- pre-fetch is out of the question.
- imagePullSecrets: can anyone publish his working yamls? 
below is my setup, can anybody spot something wrong?
** my setup **
I am distributing over aws + coreos (alpha-v774.0.0):
<secret.yaml>
apiVersion: v1
kind: Secret
metadata:
  name: myregistrykey
  namespace: staging
data:
  .dockercfg: ${MY_BASE64}
type: kubernetes.io/dockercfg
------------------------------------------------
<dbRC.yaml>
apiVersion: v1
kind: ReplicationController
metadata:
  name: db
  namespace: staging
  labels:
    name: db
spec:
  replicas: 2
  selector:
    name: db
  template:
    metadata:
      labels:
         name: db
    spec:
      imagePullSecrets:
      - name: myregistrykey
      containers:
      - name: db
        image: privateRepo/someTag:latest
------------------------------------------------

**Comment by deads2k** [comment]
> can anyone publish his working yamls? below is my setup, can anybody spot something wrong?

Can you confirm that your base64 secret matches correctly using the command referenced here: https://github.com/kubernetes/kubernetes/issues/499#issuecomment-131162957 ?  The imagePullSecret is loaded into a keyring that matches based on the URL of your pull spec, so that needs to match correctly as well.

**Comment by drora** [comment]
apparently it was a base64 issue. thanks for the lead @deads2k.
when I compered the original file to the stored one as suggested I had noticed that another jsonobject called "auth" was encapsulating my .dockerconfig file json structure from the top level.
as it turns out (I could be stating the obvious here but what the hell, it could help somebody), base64 relies on local encoding that varies between different linux distros (e.g. UTF8, UTF16, etc).
here's what fixed it for me: 
1) remove al whitespaces from your .dockerconfig file (result is a 1 liner json)
2) use the same os to encode your .dockercfg file as in your target kubelet host.

**Comment by iameli** [comment]
Hmm. Same problem, but I'm pretty confident it's not a base64 problem, as pulling the secret out of the thing works great.

```

**Comment by iameli** [comment]
I asked [a StackOverflow question](http://stackoverflow.com/questions/32510310/kubernetes-imagepullsecrets-not-working-getting-image-not-found) with largely the same problem as @drora -- cross-referencing here because I used this thread to troubleshoot.

**Comment by streamnsight** [comment]
got this working when:
1) I remove the 'auths' wrapping on the object in the config.json
2) add 'https://' in front of the URL
3) make it one line (not sure this is so critical but it seems to cause issues (invalid json) if not in one line

-> base64 encode and make sure it is still one line when listin as entry for .dockercfg

**Comment by plamer** [comment]
@streamnsight thank you! I was going nuts with this util I saw your directions :)

**Comment by ntquyen** [comment]
@streamnsight Thanks a lot! Your solution saves my day!

**Comment by chesleybrown** [comment]
Thank you @streamnsight. I think I had to remove _all_ spaces from it as well.

**Comment by streamnsight** [comment]
I pushed this PR https://github.com/kubernetes/kubernetes/pull/17286 for the documentation over a month ago but it's still has not made it into the repo.
@erictune @brendandburns @smarterclayton
